File: searchlib/src/main/java/com/yahoo/searchlib/expression/AndPredicateNode.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.searchlib.expression;

import com.yahoo.vespa.objects.Deserializer;
import com.yahoo.vespa.objects.ObjectVisitor;
import com.yahoo.vespa.objects.Serializer;

import java.util.List;

/**
 * Protocol class for {@code OrPredicate}.
 *
 * @author johsol
 */
public class AndPredicateNode extends MultiArgPredicateNode {

    public static final int classId = registerClass(0x4000 + 177, AndPredicateNode.class, AndPredicateNode::new);

    public AndPredicateNode() {}

    public AndPredicateNode(List<FilterExpressionNode> args) {
        super(args);
    }

    @Override protected int onGetClassId() { return classId; }

    @Override
    public FilterExpressionNode clone() {
        return new AndPredicateNode(getArgs().map(list -> list.stream()
                .map(FilterExpressionNode::clone).toList()).orElse(null));
    }

    @Override
    protected void onSerialize(Serializer buf) {
        super.onSerialize(buf);
    }

    @Override
    protected void onDeserialize(Deserializer buf) {
        super.onDeserialize(buf);
    }

    @Override
    public void visitMembers(ObjectVisitor visitor) {
        super.visitMembers(visitor);
    }

    @Override
    public boolean equals(Object o) {
        return super.equals(o);
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }
}

File: searchlib/src/vespa/searchlib/expression/and_predicate_node.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
#include "and_predicate_node.h"
#include "resultnode.h"

#include <algorithm>

namespace search::expression {
using vespalib::Deserializer;
using vespalib::Serializer;

IMPLEMENT_IDENTIFIABLE_NS2(search, expression, AndPredicateNode, MultiArgPredicateNode);

AndPredicateNode::AndPredicateNode() noexcept = default;

AndPredicateNode::~AndPredicateNode() = default;

AndPredicateNode::AndPredicateNode(const AndPredicateNode&) = default;

AndPredicateNode& AndPredicateNode::operator=(const AndPredicateNode&) = default;

bool AndPredicateNode::allow(const DocId docId, const HitRank rank) {
    return std::ranges::all_of(args(), [docId, rank](auto& arg){ return arg->allow(docId, rank); });
}

bool AndPredicateNode::allow(const document::Document& doc, const HitRank rank) {
    return std::ranges::all_of(args(), [&doc, rank](auto& arg){ return arg->allow(doc, rank); });
}
} // namespace search::expression
